lib/ASTGen/Sources/ASTGen/Macros.swift

```diff
@@ -664,16 +665,35 @@ private func findSyntaxNodeInSourceFile<Node: SyntaxProtocol>(
   }
 
   var currentSyntax = Syntax(token)
+  var resultSyntax: Node? = nil
   while let parentSyntax = currentSyntax.parent {
     if let typedParent = parentSyntax.as(type) {
-      return typedParent
+      resultSyntax = typedParent
+      break
     }
 
     currentSyntax = parentSyntax
   }
 
-  print("unable to find node: \(token.debugDescription)")
-  return nil
+  // If we didn't find anything, complain and fail.
+  guard var resultSyntax else {
+    print("unable to find node: \(token.debugDescription)")
+    return nil
+  }
+
+  // If we want the outermost node, keep looking.
+  if wantOutermost {
+    while let parentSyntax = resultSyntax.parent {
+      guard let typedParent = parentSyntax.as(type),
+            typedParent.position == resultSyntax.position else {
+        break;
```

Comment:
No semicolon in Swift 😉

---

Also, I’m not entirely sure I understand what `wantOutermost` does. It finds the outermost node that starts at the same location as the innermost node that contains the given position? But then, if you’ve got 
`MyType<MyGenericParam>` and you start looking for the outermost type from `MyGenericParam`, you will get `MyGenericParam` but intuitively I would imagine you to want `MyType<MyGenericParam>`

Reply:
It's very much focused on the C++ parser interleaving use case, where we want the syntax node that starts at the given location and is a type. The normal behavior of this function finds the innermost syntax node, which is generally why one wants. This one needs to walk further up the tree because if we are at `Swift.Int`, we don't want the `Swift` node, we want the whole `Swift.Int` node. Perhaps there's a better name for this.

Reply:
But if we start at `Int`, do we want `Swift.Int` or `Int`?

Reply:
We want `Int`, and that's what we would get because the start position of `Int` and the position of `Swift.Int` are different.